presto-spark-base/src/main/java/com/facebook/presto/spark/util/PrestoSparkStatsCollectionUtils.java

```diff
@@ -0,0 +1,157 @@
+        incSparkInternalAccumulator(sparkInternalAccumulatorKey, prestoKey, metricVal);
+    }
+
+    static String getSparkInternalAccumulatorKey(final String prestoKey)
+    {
+        if (prestoKey.contains(SPARK_INTERNAL_ACCUMULATOR_PREFIX)) {
+            if (prestoKey.contains(PRESTO_NATIVE_OPERATOR_STATS_SEP)) {
+                String[] strs = prestoKey.split(PRESTO_NATIVE_OPERATOR_STATS_SEP);
+                if (strs == null || strs.length < 2) {
+                    log.info("Fail to build spark internal key for %s format not supported", prestoKey);
+                    return "";
+                }
+                return strs[1];
+            }
+        }
+        String[] strs = prestoKey.split("\\.");
+        if (strs == null || strs.length < 2) {
+            log.info("Fail to build spark internal key for %s format not supported", prestoKey);
+            return "";
+        }
+        String prestoNewKey = String.format("%1$s%2$s", strs[0], strs[strs.length - 1]);
+        if (prestoNewKey.contains("_")) {
+            prestoNewKey = CaseUtils.toCamelCase(prestoKey, false, '_');
+        }
+        return String.format("%1$s%2$s%3$s", SPARK_INTERNAL_ACCUMULATOR_PREFIX,
+                PRESTO_NATIVE_OPERATOR_STATS_PREFIX, prestoNewKey);
+    }
+
+    static Set<RuntimeStats> collectRuntimeStats(TaskInfo taskInfo)
+    {
+        Set<RuntimeStats> stats = new HashSet<>();
+        if (taskInfo.getStats() == null) {
+            return stats;
+        }
+        for (PipelineStats pipelineStats : taskInfo.getStats().getPipelines()) {
+            if (pipelineStats != null) {
+                for (OperatorStats operatorStats : pipelineStats.getOperatorSummaries()) {
+                    if (operatorStats != null) {
+                        if (operatorStats.getOperatorType().equals("NativeExecutionOperator")) {
+                            NativeExecutionInfo nativeExecutionInfo = (NativeExecutionInfo) operatorStats.getInfo();
+                            if (nativeExecutionInfo != null) {
+                                for (TaskStats taskStats : nativeExecutionInfo.getTaskStats()) {
+                                    if (taskStats != null) {
+                                        RuntimeStats runtimeStat = taskStats.getRuntimeStats();
+                                        stats.add(runtimeStat);
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        return stats;
+    }
+
+    static Long getMetricValue(RuntimeMetric metric, boolean isSparkUnitMs)
+    {
+        BigDecimal sum = BigDecimal.valueOf(metric.getSum());
+        if (metric.getUnit().equals(RuntimeUnit.NANO) && isSparkUnitMs) {
+            sum = sum.divide(BigDecimal.valueOf(1000000), BigDecimal.ROUND_DOWN);
+        }
+        BigDecimal count = BigDecimal.valueOf(metric.getCount());
```

Comment:
Why is the count is in `Decimal`? Shouldn't it be `long`? BTW, do we need to aggregate the value by ourself? I'm under the impression that the accumulator will do the aggregation.

Reply:
I am using BigDecimal to avoid data overflow during division. count is in BigDecimal because  both numerator and the denominator have to be Bigdecimal .

Rational for computing the average:
In this use case we are collecting taskMetrics only once on task completion.
Let's say we collected n values while the task was running. 
Presto  RuntimesStat will store sum, min, max and count
 We are adding sum/count to the accumulator  which is the same as accumulator.value() if the n values were collected directly through the accumulator.

Reply:
If the value we get from `metric.getCount()`  is a valid long, in theory we won't get overflow right?

For computing the average, 
we usually use accumulator.sum to get and log the values from TaskMetrics, so it'd be better to keep the raw sum value otherwise the number will be unexpected.

Reply:
Ok, may be I am being overly cautious , I think we can still have overflow.  the average can be a decimal even if you start with long. That is if you want an accurate number.  For instance, the avg for 8, 5, 7 => is a decimal. 

Regarding the average:
Let 's consider a stage with  with 2 tasks. 
task1:   2 3 4 5  sum is 14 and count is 4
task2 :  14        sum is  14  and count is 1
avg for stage => is  28/5 if I only save sum  the avg at stage level will be 28/2 which is wrong
However, as you mentioned, if each aggregation level only store sum then computing avg is not needed.

Reply:
Yeah, for current approach we can only store either avg or sum (if you store avg you can't get sum any more). By looking at the taskMetrics in fb-spark, we mostly populate sum from the taskMetrics, so I believe storing sum directly here can cover most of the cases.

Reply:
I switched to sum . Cosco metrics are mostly sum. This can lead to some strange metrics in spark_app table but the problem won't be unique to Prestospark.